File: src/gesture/HandGestureController.ts
import { CursorSettings, Gesture, NormalizedLandmark } from '../types';
import { GestureAnalyzer } from './GestureAnalyzer';
import { ProximityDetector } from './ProximityDetector';
import { PhoneActionExecutor } from '../overlay/PhoneActionExecutor';

export class HandGestureController {
  private isAwake = false;
  private gestureAnalyzer = new GestureAnalyzer();
  private actionExecutor = new PhoneActionExecutor();
  private proximityDetector: ProximityDetector;
  private lastExecutedGesture = Gesture.NONE;
  private lastGestureTime = 0;
  private readonly GESTURE_THROTTLE_MS = 600; // Prevent rapid firing of same gesture

  constructor() {
    this.proximityDetector = new ProximityDetector(
      () => this.enterActiveMode(),
      () => this.enterSleepMode()
    );
  }

  getActionExecutor(): PhoneActionExecutor {
    return this.actionExecutor;
  }

  getGestureAnalyzer(): GestureAnalyzer {
    return this.gestureAnalyzer;
  }

  processLandmarks(landmarks: NormalizedLandmark[] | null): void {
    if (!landmarks || landmarks.length === 0) {
      this.proximityDetector.checkSleep();
      return;
    }

    // Hand proximate check
    this.proximityDetector.checkProximity(landmarks);

    if (!this.isAwake) {
      return;
    }

    // Always update cursor position if index finger (landmark 8) is detected
    if (landmarks.length > 8) {
      const indexFinger = landmarks[8];
      // Invert X because the front camera is mirrored (1 - x), matching Android source code
      this.actionExecutor.updateCursor(1 - indexFinger.x, indexFinger.y);
    }

    const gesture = this.gestureAnalyzer.analyze(landmarks);
    const now = Date.now();

    if (gesture !== Gesture.NONE) {
      // For continuous gestures like CURSOR_MOVE or repeated scroll/volume, or discrete gestures with throttle
      const isContinuous = gesture === Gesture.CURSOR_MOVE;
      const canExecute = isContinuous || (now - this.lastGestureTime > this.GESTURE_THROTTLE_MS) || (gesture !== this.lastExecutedGesture);

      if (canExecute) {
        this.actionExecutor.execute(gesture);
        this.lastExecutedGesture = gesture;
        this.lastGestureTime = now;
      }
    }
  }

  enterActiveMode(): void {
    this.isAwake = true;
    this.actionExecutor.showWakeOverlay();
  }

  enterSleepMode(): void {
    this.isAwake = false;
    this.actionExecutor.hideOverlays();
  }

  getIsAwake(): boolean {
    return this.isAwake;
  }

  updateSettings(settings: CursorSettings): void {
    this.actionExecutor.updateSettings(settings);
  }

  release(): void {
    this.enterSleepMode();
    this.proximityDetector.reset();
    this.gestureAnalyzer.reset();
  }
}

File: android/app/src/main/java/com/gesture/controller/gesture/HandGestureController.kt
package com.gesture.controller.gesture

import com.gesture.controller.overlay.CursorSettings
import com.gesture.controller.overlay.PhoneActionExecutor

class HandGestureController {
    private var isAwake = false
    private val gestureAnalyzer = GestureAnalyzer()
    private val actionExecutor = PhoneActionExecutor()
    private val proximityDetector: ProximityDetector
    private var lastExecutedGesture = Gesture.NONE
    private var lastGestureTime = 0L
    private val gestureThrottleMs = 600L

    init {
        proximityDetector = ProximityDetector(
            onWakeDetected = { enterActiveMode() },
            onSleepDetected = { enterSleepMode() }
        )
    }

    fun getActionExecutor(): PhoneActionExecutor = actionExecutor

    fun getGestureAnalyzer(): GestureAnalyzer = gestureAnalyzer

    fun processLandmarks(landmarks: List<NormalizedLandmark>?) {
        if (landmarks == null || landmarks.isEmpty()) {
            proximityDetector.checkSleep()
            return
        }

        proximityDetector.checkProximity(landmarks)

        if (!isAwake) {
            return
        }

        // Steer cursor with index finger tip (landmark index 8)
        if (landmarks.size > 8) {
            val indexFinger = landmarks[8]
            // Invert X because front camera is mirrored (1f - x)
            actionExecutor.updateCursor(1f - indexFinger.x, indexFinger.y)
        }

        val gesture = gestureAnalyzer.analyze(landmarks)
        val now = System.currentTimeMillis()

        if (gesture != Gesture.NONE) {
            val isContinuous = gesture == Gesture.CURSOR_MOVE
            val canExecute = isContinuous || (now - lastGestureTime > gestureThrottleMs) || (gesture != lastExecutedGesture)

            if (canExecute) {
                actionExecutor.execute(gesture)
                lastExecutedGesture = gesture
                lastGestureTime = now
            }
        }
    }

    fun enterActiveMode() {
        isAwake = true
        actionExecutor.showWakeOverlay()
    }

    fun enterSleepMode() {
        isAwake = false
        actionExecutor.hideOverlays()
    }

    fun getIsAwake(): Boolean = isAwake

    fun updateSettings(settings: CursorSettings) {
        actionExecutor.updateSettings(settings)
    }

    fun release() {
        enterSleepMode()
        proximityDetector.reset()
        gestureAnalyzer.reset()
    }
}
